File: graphs/java/hasPath.java
package linear_concepts.graphs.java;

import java.util.*;

public class hasPath {
    public static void main(String[] args) {
        Map<String, List<String>> graph = new HashMap<>();
        graph.put("f", Arrays.asList("g", "i"));
        graph.put("g", Arrays.asList("h"));
        graph.put("h", new ArrayList<>());
        graph.put("i", Arrays.asList("g","k"));
        graph.put("j", Arrays.asList("i"));
        graph.put("k", new ArrayList<>());

        System.out.println(hasPath_DFS(graph, "f", "k"));

        System.out.println("---------------------------Recursive---------------------------------");
        System.out.println(hasPath_DFS_Recursive(graph, "f","k"));
        System.out.println("---------------------------BFS---------------------------------");
        System.out.println(hasPath_BFS(graph, "f","k"));


    }

    public static boolean hasPath_DFS(Map<String, List<String>> graph, String source, String Dest) {

        Stack<String> stack = new Stack<String>();
        stack.push(source);

        while(!stack.isEmpty()){
            String current = stack.pop();
            if(current.equals(Dest))
                return true;
            
            List<String> neighbors = graph.get(current);

            for (String neighbor: neighbors){
                stack.push(neighbor);
            }
        }
        return false;
        
    }
    
    public static boolean hasPath_DFS_Recursive(Map<String, List<String>> graph, String source, String Dest){
        if (source.equals(Dest))
            return true;
        for(String neighbor: graph.get(source)){
            if(hasPath_DFS_Recursive(graph, neighbor, Dest))
                return true;
        }
        return false;
    }
    public static boolean hasPath_BFS(Map<String, List<String>> graph, String source, String Dest) {

        Queue<String> queue = new LinkedList<String>();
        queue.offer(source);
        while(!queue.isEmpty()){
            String current = queue.poll();
            if(current.equals(Dest))
                return true;
            List<String> neighbors = graph.get(current);
            for(String neighbor: neighbors){
                queue.offer(neighbor);
            }
        }
        return false;
    }

}

File: graphs/js/hasPath.js
const graph = {
    f: ['g','i'],
    g: ['h'],
    h:[],
    i:['g','k'],
    j:['i'],
    k:[]
}

const hasPath_DFS = (graph, source, dest) =>{
    
    const stack = [source]

    while(stack.length > 0)
    {
        const current = stack.pop()
         
        if(current == dest)
            return true

        for (let neighbor of graph[current]){
            stack.push(neighbor)
        }
    }
    return false

}

const hasPath_DFS_Recursive = (graph,source,dest) => {
    if (source === dest)
        return true

    for( let neighbor of graph[source]){
        if(hasPath_DFS_Recursive(graph, neighbor, dest))
            return true
    }
    return false
}

const hasPath_BFS = (graph, source, dest) =>{
    
    const queue = [source]

    while(queue.length > 0){
        const current = queue.shift()
        if (current === dest)
            return true
        for (let neighbor of graph[current]){
            queue.push(neighbor)
        }
    }
    return false

}

console.log(hasPath_DFS(graph,'f','k'))
console.log(hasPath_DFS_Recursive(graph,'f','k'))
console.log(hasPath_BFS(graph,'f','k'))
